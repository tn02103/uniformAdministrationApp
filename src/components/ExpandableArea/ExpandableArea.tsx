import { faChevronDown } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { useState } from "react";
import { Row, Button } from "react-bootstrap";
import styles from "./ExpandableArea.module.css";
import { useI18n } from "@/lib/locales/client";

export const ExpandableArea = ({ children }: { children: React.ReactNode }) => {
    const t = useI18n();
    const [expanded, setExpanded] = useState(false);

    return (
        <>
            <Row style={{ height: "2rem" }} className="align-items-center justify-content-center">
                <div className="position-relative">
                    <hr className="" />
                    <div
                        className="w-auto bg-white align-center position-absolute top-50 start-50 translate-middle"
                    >
                        <Button
                            size="sm"
                            variant="light"
                            className="border-0"
                            onClick={() => setExpanded(!expanded)}
                        >
                            <FontAwesomeIcon icon={faChevronDown} className={"text-dark me-2 " + (expanded && styles.open)} />
                            {expanded ? t('expandableArea.showLess') : t('expandableArea.showMore')}
                        </Button>
                    </div>
                </div>
            </Row>
            {expanded && (
                <Row>
                    {children}
                </Row>
            )}
        </>
    );
}
