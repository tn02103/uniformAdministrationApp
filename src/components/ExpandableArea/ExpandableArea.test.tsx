import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { ExpandableArea } from "./ExpandableArea";

describe('ExpandableArea', () => {

    it('renders area closed', () => {
        const { container } = render(
            <ExpandableArea>
                <div>Test</div>
            </ExpandableArea>
        );

        const button = screen.getByRole('button');
        expect(button).toHaveTextContent('expandableArea.showMore');
        expect(screen.getByRole('img', { hidden: true })).not.toHaveClass(/open/);

        expect(screen.queryByText('Test')).toBeNull();
        expect(container).toMatchSnapshot();
    });

    it('renders area open', async () => {
        const user = userEvent.setup();
        const { container } = render(
            <ExpandableArea>
                <div>Test</div>
            </ExpandableArea>
        );
        const button = screen.getByRole('button');

        await user.click(button);
        expect(button).toHaveTextContent('expandableArea.showLess');
        expect(screen.getByRole('img', { hidden: true })).toHaveClass(/open/);

        expect(screen.getByText('Test')).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });
});