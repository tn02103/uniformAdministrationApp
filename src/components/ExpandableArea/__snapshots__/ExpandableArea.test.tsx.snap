// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ExpandableArea renders area closed 1`] = `
<div>
  <div
    class="align-items-center justify-content-center row"
    style="height: 2rem;"
  >
    <div
      class="position-relative"
    >
      <hr
        class=""
      />
      <div
        class="w-auto bg-white align-center position-absolute top-50 start-50 translate-middle"
      >
        <button
          class="border-0 btn btn-light btn-sm"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-chevron-down text-dark me-2 false"
            data-icon="chevron-down"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M201.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 338.7 54.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
              fill="currentColor"
            />
          </svg>
          expandableArea.showMore
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ExpandableArea renders area open 1`] = `
<div>
  <div
    class="align-items-center justify-content-center row"
    style="height: 2rem;"
  >
    <div
      class="position-relative"
    >
      <hr
        class=""
      />
      <div
        class="w-auto bg-white align-center position-absolute top-50 start-50 translate-middle"
      >
        <button
          class="border-0 btn btn-light btn-sm"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-chevron-down text-dark me-2 open"
            data-icon="chevron-down"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M201.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 338.7 54.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
              fill="currentColor"
            />
          </svg>
          expandableArea.showLess
        </button>
      </div>
    </div>
  </div>
  <div
    class="row"
  >
    <div>
      Test
    </div>
  </div>
</div>
`;
